lvgl_tft/display_port.h

```diff
@@ -15,6 +15,17 @@ extern "C"
 #include <stdint.h>
 #include <stdbool.h>
 
+typedef enum {
+    CMD_WIDTH_8BITS,
+    CMD_WIDTH_16BITS,
+    CMD_WIDTH_INVALID,
+} cmd_width_t;
+
+typedef enum {
+    DATA_XFER_MODE_SYNC,
+    DATA_XFER_MODE_ASYNC,
+} data_xfer_mode_t;
```

Comment:
Why do we need to differentiate sync and async transfer at this level? Why can't the platform-specific driver decide how to send?

Reply:
Removed

Reply:
Thanks! 

One more thing came to my mind: the prototype should be something like:
```c
void display_interface_update_image(lv_disp_drv_t *drv, lv_arae_t * area, lv_color_t * data);
```

It's because in case of devices with pure frame buffer(s), without a display controller, we can't send commands to set the "window". Instead they want to swap the frame buffers, memcpy/DMA the rendered image, etc.

So this command should tell the area too.

Reply:
I think I understand you now, what if we add another display port API for those kind of displays instead of trying to have generic one?

Let me try to confirm my understanding with an example.

We have N display types:
- An ILI9341
- Pure frame buffer

With the ILI9341 we need to let the display know the area coordinates with commands and then send the color data, like so:
```c
void ili9341_flush(lv_disp_drv_t * drv, const lv_area_t * area, lv_color_t * color_map)
{
    uint8_t data[4] = {0};
    uint32_t size = lv_area_get_width(area) * lv_area_get_height(area);

    /*Column addresses*/
    data[0] = (area->x1 >> 8) & 0xFF;
    data[1] = area->x1 & 0xFF;
    data[2] = (area->x2 >> 8) & 0xFF;
    data[3] = area->x2 & 0xFF;
    
    display_interface_send_cmd(drv, 0x2A, CMD_WIDTH_8BITS, data, sizeof(data));

    /* Page addresses */
    data[0] = (area->y1 >> 8) & 0xFF;
    data[1] = area->y1 & 0xFF;
    data[2] = (area->y2 >> 8) & 0xFF;
    data[3] = area->y2 & 0xFF;
    
    display_interface_send_cmd(drv, 0x2B, CMD_WIDTH_8BITS, data, sizeof(data));

    /* Memory write */
    display_interface_send_cmd(drv, 0x2C, CMD_WIDTH_8BITS, NULL, 0);
    display_interface_send_data(drv, color_map, size * 2);
}
```

With the pure frame buffer display:
```c
void pure_frame_buffer_flush(lv_disp_drv_t * drv, const lv_area_t * area, lv_color_t * color_map)
{
    display_interface_send_frame_buffer(drv, area, color_map); // equivalent to your suggestion
}
```

---

If we make a generic function, the flush callbacks would end up being just wrappers of it.
```c
void ili9341_flush(lv_disp_drv_t * drv, const lv_area_t * area, lv_color_t * color_map)
{
    display_interface_update_image(drv, area, color_map);
}
```

```c
void pure_frame_buffer_flush(lv_disp_drv_t * drv, const lv_area_t * area, lv_color_t * color_map)
{
    display_interface_update_image(drv, area, color_map);
}
```

What do you think?

Reply:
I vote for generic, because... it's more generic :laughing: I think it's better to have one function instead of two similar. The users can ignore the `area` parameter if they don't use it.

Reply:
I'm not totally convinced to be honest, I think the `display_interface_update_image` function would end up having a lot of code inside. Can we wait for more opinions? What do you think?

Reply:
>I think the display_interface_update_image function would end up having a lot of code inside.  

Why will it have more code than `display_interface_send_data`?

Reply:
So, this would be the flush callback of the ssd1306 display driver:

```c
void ssd1306_flush(lv_disp_drv_t * disp_drv, const lv_area_t * area, lv_color_t * color_p)
{
    /* Divide by 8 */
    uint8_t row1 = area->y1 >> 3;
    uint8_t row2 = area->y2 >> 3;

    uint8_t conf[] = {
        OLED_CONTROL_BYTE_CMD_STREAM,
        OLED_CMD_SET_MEMORY_ADDR_MODE,
        0x00,
        OLED_CMD_SET_COLUMN_RANGE,
        (uint8_t) area->x1,
        (uint8_t) area->x2,
        OLED_CMD_SET_PAGE_RANGE,
        row1,
        row2,
    };

    display_interface_send_cmd(disp_drv, OLED_CONTROL_BYTE_CMD_STREAM, &conf[1], sizeof(conf) - 1);
    display_interface_send_data(disp_drv, color_p, OLED_COLUMNS * (1 + row2 - row1));
    lv_disp_flush_ready(disp_drv);
}
```

If we use the `display_interface_send_image` interface we would need to handle the specifics of ssd1306 inside, rather than in the flush callback:

```c
void ssd1306_flush(lv_disp_drv_t * disp_drv, const lv_area_t * area, lv_color_t * color_p)
{
    display_interface_send_image(drv, area, color_p);
}

void display_interface_send_image(lv_disp_drv_t *drv, lv_area_t *area, lv_color_t *color)
{
    if (display_is_ssd1306) {
    /* Divide by 8 */
    uint8_t row1 = area->y1 >> 3;
    uint8_t row2 = area->y2 >> 3;

    uint8_t conf[] = {
        OLED_CONTROL_BYTE_CMD_STREAM,
        OLED_CMD_SET_MEMORY_ADDR_MODE,
        0x00,
        OLED_CMD_SET_COLUMN_RANGE,
        (uint8_t) area->x1,
        (uint8_t) area->x2,
        OLED_CMD_SET_PAGE_RANGE,
        row1,
        row2,
    };

    display_interface_send_cmd(disp_drv, OLED_CONTROL_BYTE_CMD_STREAM, &conf[1], sizeof(conf) - 1);
    display_interface_send_data(disp_drv, color_p, OLED_COLUMNS * (1 + row2 - row1));

    lv_disp_flush_ready(disp_drv);
    }
}
```

Reply:
I meant the end of the `ssd1306_flush` looks like this:

```c
    display_interface_send_cmd(disp_drv, OLED_CONTROL_BYTE_CMD_STREAM, &conf[1], sizeof(conf) - 1);
    display_interface_send_image(disp_drv, area, color_p);
```

So you give the "window commands" and then ask to send the image too. Finally, only `display_interface_send_data` needs to be called. With "pure frame buffer" display `display_interface_send_cmd` won't be called only `display_interface_send_image`.

Reply:
Do you mean to implement something like this?

```c
void display_interface_send_image(lv_disp_drv_t * drv, const lv_area_t * area, lv_color_t * color_map)
{
    if (is_pure_frame_buffer(drv)) {
        pure_frame_buffer_flush(drv, area, color_map);
    }
    else {
        display_interface_send_data(drv, color_map, /* Calculate bytes to send from area */);
    }
}
```

If not, would you care to have a sketch of how `display_interface_send_image` would look inside?

Reply:
Yes, exactly. To complete your example:
```c
uint32_t s = lv_area_get_size(area) * sizeof(lv_color_t);
display_interface_send_data(drv, color_map, s);
```

What do you think?

(I'd call it `display_interface_update_image` as it's not sure it send anything, but it's a question detail)

Reply:
Is there a configuration value we can look for to know if a driver is a pure frame buffer display?

Reply:
Not really, only its flush_cb does its job differently.